src/v/datalake/tests/record_multiplexer_bench.cc

```diff
@@ -0,0 +1,564 @@
+    static constexpr ::testing::protobuf_generator_config gen_config = {
+      .string_length_range{302, 302}};
+
+    static std::string proto_schema = generate_linear_proto(1);
+
+    static thread_local chunked_vector<model::record_batch> batch_data
+      = co_await generate_protobuf_batches(
+        records_per_batch,
+        batches,
+        "proto_schema",
+        proto_schema,
+        {0},
+        gen_config);
+
+    auto reader = model::make_fragmented_memory_record_batch_reader(
+      share_batches(batch_data));
+
+    auto consumer = counting_consumer{.mux = create_mux()};
+
+    perf_tests::start_measuring_time();
+    auto res = co_await reader.consume(std::move(consumer), model::no_timeout);
+    perf_tests::stop_measuring_time();
+
+    co_return res.total_bytes;
+}
+
+PERF_TEST_CN(
+  record_multiplexer_bench_fixture,
+  protobuf_381_byte_message_linear_40_fields) {
+    static constexpr size_t records_per_batch = 10;
+    static constexpr size_t batches = 1000;
+    static constexpr ::testing::protobuf_generator_config gen_config = {
+      .string_length_range{5, 5}};
+
+    // Generate a schema near the nesting limit.
+    // (12 payload + 3 header) bytes per field
+    static std::string proto_schema = generate_linear_proto(40);
+
+    static thread_local chunked_vector<model::record_batch> batch_data
+      = co_await generate_protobuf_batches(
+        records_per_batch,
+        batches,
+        "proto_schema",
+        proto_schema,
+        {0},
+        gen_config);
+
+    auto reader = model::make_fragmented_memory_record_batch_reader(
+      share_batches(batch_data));
+
+    auto consumer = counting_consumer{.mux = create_mux()};
+
+    perf_tests::start_measuring_time();
+    auto res = co_await reader.consume(std::move(consumer), model::no_timeout);
+    perf_tests::stop_measuring_time();
+
+    co_return res.total_bytes;
+}
+
+PERF_TEST_CN(
+  record_multiplexer_bench_fixture,
+  protobuf_381_byte_message_linear_80_fields) {
+    static constexpr size_t records_per_batch = 10;
+    static constexpr size_t batches = 1000;
+    static constexpr ::testing::protobuf_generator_config gen_config = {
+      .string_length_range{1, 1}};
+
+    // Generate a schema near the nesting limit.
+    // (3 payload + 3 header) bytes per field
+    static std::string proto_schema = generate_linear_proto(80);
+
+    static thread_local chunked_vector<model::record_batch> batch_data
+      = co_await generate_protobuf_batches(
+        records_per_batch,
+        batches,
+        "proto_schema",
+        proto_schema,
+        {0},
+        gen_config);
+
+    auto reader = model::make_fragmented_memory_record_batch_reader(
+      share_batches(batch_data));
+
+    auto consumer = counting_consumer{.mux = create_mux()};
+
+    perf_tests::start_measuring_time();
+    auto res = co_await reader.consume(std::move(consumer), model::no_timeout);
+    perf_tests::stop_measuring_time();
+
+    co_return res.total_bytes;
```

Comment:
nit: I'm wondering if we should wrap this in some `measure_proto(std::string schema)` and use it everywhere (same for avro)

Reply:
It'd be possible, I originally avoided it so that folks could look at a given microbenchmark and see right away what was being measured. However, it is rather repetitive. Happy to factor it out if you think it would be better that way.

Reply:
I don't particularly read microbenchmarks frequently, but it feels like it might make it'd be nice to encapsulate those details (e.g. so it's more obvious to readers that each one is actually measuring in the same way, if that's the intent. Though maybe that's expected out of a microbench? ultimately I'll leave the call up to you)

Reply:
Encapsulation it is then, don't have a strong opinion on this myself.